libc++ conflict with libstdc++ with Linux and Anaconda

When I try to run `Language.build_library('languages.so', ['tree-sitter-python'])` on a machine using an Anaconda python distribution (which uses libstdc++) and my Ubuntu 16 system has both libc++ and libstdc++ installed, `build_library` [gives higher precedence](https://github.com/tree-sitter/py-tree-sitter/blob/dd07df5434a9795e6079707cbbb4a048c4a3885a/tree_sitter/__init__.py#L47) to the existence of libc++, when libstdc++ is necessary for use with my Anaconda python. The result of this is that I get the error

`/usr/bin/ld: cannot find -lc++`.

I recognize this is strange since I have both installed, but perhaps the distutils compiler only pays attention to Anaconda library paths?

I fixed it on my system by giving higher precedence to libstdc++ (and can submit a PR), but would this break py-tree-sitter for others? Is there another way to fix this logic so it works? Since libc++ is 'newer' than libstdc++, perhaps the precedence of the older one makes sense?

> What would you consider a thorough test for this? Would passing the test suite be enough?

👍  Yes.

> Note that the build failure for PyPy on Mac is not related to the patch, but also present on current master

That's strange; I personally develop on a mac, and the tests run fine for me. The error message is interesting:

```
clang: warning: include path for libstdc++ headers not found; pass '-stdlib=libc++' on the command line to use the libc++ standard library instead [-Wstdlibcxx-not-found]
tests/fixtures/tree-sitter-python/src/scanner.cc:2:10: fatal error: 'vector' file not found
#include <vector>
         ^~~~~~~~
```

It looks like, for *some* reason, the flag `-stdlib=libstdc++` is being passed to clang, indicating that it should use the non-default C++ standard library.

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -1,0 +1,57 @@
+name: CI
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+
+jobs:
+  build:
+    strategy:
+      fail-fast: false
+      matrix:
+        os:
+          - ubuntu-latest
+          - macos-latest
+          - windows-latest
+        python:
+          - "3.5"
+          - "3.8"
+          - pypy-3.7
+        install_libcxx:
+          - false
+          - true
+        exclude:
+          - os: windows-latest
+            install_libcxx: true
+          - os: macos-latest
+            install_libcxx: true
+          - os: macos-latest
+            python: pypy-3.7
+    runs-on: ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          submodules: true
+      - uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python }}
+      - name: Uninstall libcxx
+        if: ${{ matrix.os == 'ubuntu-latest' && !matrix.install_libcxx }}
+        run: |
+          DEBIAN_FRONTEND=noninteractive sudo apt-get purge -y 'libc++*'
+      - name: Install libcxx
+        if: ${{ matrix.install_libcxx }}
+        run: |
+          DEBIAN_FRONTEND=noninteractive sudo apt-get install -y libc++-dev libc++abi-dev
+      - name: Install flake8
+        run: python -m pip install flake8
+      - name: Fetch fixtures
+        run: ./script/fetch-fixtures
+      - name: Lint
+        run: ./script/lint
+      - name: Test
+        shell: bash
+        run: |
+          CFLAGS="-O0 -g" python setup.py test

--- a/tree_sitter/__init__.py
+++ b/tree_sitter/__init__.py
@@ -1,8 +1,8 @@
 """Python bindings for tree-sitter."""
 
 from ctypes import cdll, c_void_p
-from ctypes.util import find_library
 from distutils.ccompiler import new_compiler
+from distutils.unixccompiler import UnixCCompiler
 from os import path
 from platform import system
 from tempfile import TemporaryDirectory
@@ -43,11 +43,8 @@ class Language:
         ]
 
         compiler = new_compiler()
-        if cpp:
-            if find_library("c++"):
-                compiler.add_library("c++")
-            elif find_library("stdc++"):
-                compiler.add_library("stdc++")
+        if isinstance(compiler, UnixCCompiler):
+            compiler.compiler_cxx[0] = "c++"
 
         if max(source_mtimes) <= output_mtime:
             return False
@@ -69,7 +66,11 @@ class Language:
                         extra_preargs=flags,
                     )[0]
                 )
-            compiler.link_shared_object(object_paths, output_path)
+            compiler.link_shared_object(
+                object_paths,
+                output_path,
+                target_lang="c++" if cpp else "c",
+            )
         return True
 
     def __init__(self, library_path, name):

